src/emerge.cpp

```diff
@@ -524,21 +537,39 @@ bool EmergeThread::popBlockEmerge(v3s16 *pos, BlockEmergeData *bedata)
 }
 
 
-EmergeAction EmergeThread::getBlockOrStartGen(
-	const v3s16 &pos, bool allow_gen, MapBlock **block, BlockMakeData *bmdata)
+EmergeAction EmergeThread::getBlockOrStartGen(const v3s16 pos, bool allow_gen,
+	 const std::string *from_db, MapBlock **block, BlockMakeData *bmdata)
 {
 	MutexAutoLock envlock(m_server->m_env_mutex);
 
+	auto block_ok = [] (MapBlock *b) {
+		return b && b->isGenerated();
+	};
+
 	// 1). Attempt to fetch block from memory
 	*block = m_map->getBlockNoCreateNoEx(pos);
 	if (*block) {
-		if ((*block)->isGenerated())
+		if (block_ok(*block))
 			return EMERGE_FROM_MEMORY;
 	} else {
-		// 2). Attempt to load block from disk if it was not in the memory
-		*block = m_map->loadBlock(pos);
-		if (*block && (*block)->isGenerated())
+		if (!from_db) {
+			// 2). We should attempt loading it
 			return EMERGE_FROM_DISK;
+		}
+		// 2). Second invocation, we have the data
+		if (!from_db->empty()) {
+			*block = m_map->getBlockNoCreateNoEx(pos);
+			if (*block) {
```

Comment:
This is (almost?) the same as at `1).`, so redundant.

(The claim I made about the data race might've been wrong. I didn't check back then if block existence was already checked in this function.)

Reply:
right.
the data race in the previous revision was real.